File: Express/src/api/v1/models/Log.js
const mongoose = require("mongoose");
const toJSON = require("./plugins/toJson");

const logSchema = new mongoose.Schema(
  {
    level: {
      type: String,
      required: true,
      enum: ['info', 'warn', 'error', 'debug'], // Log levels
      default: 'info',
    },
    message: {
      type: String,
      required: true,
    },
    timestamp: {
      type: Date,
      default: Date.now,
    },
    meta: {
      type: mongoose.Schema.Types.Mixed,
      default: {},
    },
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: false,
    }
  },
  {
    timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' }
  }
);

// Apply the toJSON plugin to the log schema
logSchema.plugin(toJSON);

const Log = mongoose.model("Log", logSchema);

module.exports = Log;

File: fastapi/fastapi-rest/app/schemas/log.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime
import enum

class LogLevelEnum(str, enum.Enum):
    info = "info"
    warn = "warn"
    error = "error"
    debug = "debug"

class LogBase(BaseModel):
    level: LogLevelEnum
    message: str
    timestamp: datetime = datetime.utcnow()
    meta: Optional[str] = None

class LogCreate(LogBase):
    pass

class Log(LogBase):
    id: int
    user_id: Optional[int] = None

    class Config:
        from_attributes = True
